sdk/java-sdk-spring/src/main/scala/kalix/spring/impl/KalixSpringApplication.scala

```diff
@@ -373,16 +377,24 @@ case class KalixSpringApplication(applicationContext: ApplicationContext, config
     ReflectiveActionProvider.of(
       clz,
       messageCodec,
-      context =>
-        wiredInstance(clz) {
-          case p if p == classOf[ActionCreationContext] => context
-          case p if p == classOf[KalixClient]           => kalixClient(context)
-          case p if p == classOf[ComponentClient]       => componentClient(context)
-          case p if p == classOf[WebClientProvider]     => webClientProvider(context)
-          case p if p == classOf[Tracer] =>
-            context.getOpenTelemetryTracer.orElseGet(() =>
-              throw new IllegalStateException("Tracer not available. Make sure to have enabled tracing."))
-        })
+      context => {
+        val action =
+          wiredInstance(clz) {
+            case p if p == classOf[ActionCreationContext] => context
+            case p if p == classOf[KalixClient]           => kalixClient(context)
+            case p if p == classOf[ComponentClient]       => componentClient(context)
+            case p if p == classOf[WebClientProvider]     => webClientProvider(context)
+            case p if p == classOf[Tracer] =>
+              context.getOpenTelemetryTracer.orElseGet(() =>
+                throw new IllegalStateException("Tracer not available. Make sure to have enabled tracing."))
+          }
+
+        val field = lookupSuperclassField("kalixClient", action.getClass.getSuperclass)
```

Comment:
Here's the magic sauce. Nice.

Then how would this work in https://github.com/lightbend/kalix-jvm-sdk/pull/2022
Let me try and fail. Each time the user's Action gets instantiated it will set the `kalixClient` with the `ActionCreationContext` so when calling `componentClient()` to create a deferred call, the `ComponentCall` can pick it and ...failing here ... because we need `ActionContext` not `ActionCreationContext` to pick the metadata to add to the deferred call. What do you think?

Reply:
I'm not sure I'm following you.

To build a trace enabled ComponentClient we need the `kalixClient` and the call context (not the `ActionCreationContext`).

Maybe the code is misleading because it gives the impression that `kalixClient` needs `ActionCreationContext`. That's not true. It needs the Akka materializer only. And we get it from the creation context. Actually, that's the same for all calls.

So here, I'm inject the `kalixClient` in a field. So now, when the user call the method `componentClient()` inside the command handler we have access to the call context and the `kalixClient`. As such, we can build a fresh `ComponentClient` that is tweaked to use the tracing from the call context. 

Each call to `componentClient()` would return a new `ComponentClient` and that's what we want because it will have a state now that is bounded to the call context.

To make it work with #2022, we will need to [remove this](https://github.com/lightbend/kalix-jvm-sdk/pull/2022/files#diff-0001cb59d6b3e4b2fe7d887f42329155d35299acff67b267a37f9f5db1b77e50R61) and add a optional `ActionContext` to the `ComponentClient` constructor.

Reply:
Oh, Ok. Right, `componentClient()` it's a method now. Of course! Thanks for the explanation.